### PR TITLE
Add more backwards compatibly for MockWaveformExtractor

### PR DESCRIPTION
**Pull request by chrishalcrow**
Added several methods to MockWaveformExtractor and changed arguments in compute_blah_extensions function.

Will keep in draft until tests are added.

Added:
- format
- get_available_extension_names
- precompute_templates
- sample_spikes
- run_extract_waveforms
- select_units
- delete_extension
- get_random_spikes (new - without a way of computing some random spikes, you can't compute templates and other things)

to MockWaveformExtractor

I changed the FunctionFactory for the compute_blah_extensions (e.g. compute_principal_components, compute_templates) for backwards compatibility. Previously these took the form `compute_principal_components(waveform_extractor=we)` as well as`compute_principal_components(we)`.

Not added some methods. These involve: 
- Extension classes (extensions are now objects of a new extension class) so it's hard to see how these will not break (extensions, get_extension_class, is_extension, load_extension, register_extension)
- Loading stuff after loading your initial MockWaveformExtractor (load, load_from_folder, load_from zarr)
-Segments since SortingAnalyzer seems to deal with segments a bit differently (get_template_segment and get_waveform_segment).

### CODE REVIEW

**Comment by samuelgarcia** [comment]
Hi Chris,
thanks for this. I will have a look soon.
As a general comments the goal of backward compatibility was more load/use old computation rather than load/use/recompute old computation.

**Comment by chrishalcrow** [comment]
> Hi Chris, thanks for this. I will have a look soon. As a general comments the goal of backward compatibility was more load/use old computation rather than load/use/recompute old computation.

Sure. I was just aiming for as little broken as possible. With these changes, the pipelines in our lab work without any modification, and create lots of nice SortingAnalyzer folders as they go.

File: src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py [hunk 1]
```diff
@@ -350,6 +407,7 @@ def load_waveforms(
     if output == "SortingAnalyzer":
         return sorting_analyzer
     elif output in ("WaveformExtractor", "MockWaveformExtractor"):
+        sorting_analyzer.compute("random_spikes")
```
Comment (alejoe91) [inline src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py:410]: Remove this and add `sorting_analyzer.extensions["random_spikes"] = ext` after the `SelectRandomSpikes` is instantiated in `_read_old_waveforms_extractor_binary`

File: src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py [hunk 2]
```diff
@@ -130,6 +130,12 @@ def has_waveforms(self) -> bool:
     def delete_waveforms(self) -> None:
         self.sorting_analyzer.delete_extension("waveforms")
 
+    def delete_extension(self, extension) -> None:
+        self.sorting_analyzer.delete_extension()
+
+    def get_random_spikes(self) -> None:
+        self.sorting_analyzer.compute("random_spikes")
```
Comment (alejoe91) [inline src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py:137]: Remove

File: src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py [hunk 3]
```diff
@@ -223,17 +232,60 @@ def folder(self):
...
+        sample_spikes = dict.fromkeys(self.sorting_analyzer.unit_ids, np.array([], "int"))
+        spikes = self.sorting_analyzer.get_extension("random_spikes").some_spikes()
+        spike_indices = self.sorting_analyzer.get_extension("random_spikes").data["random_spikes_indices"]
+        for i, spike in enumerate(spikes):
+            sample_spikes[spike["unit_index"]] = np.append(sample_spikes[spike["unit_index"]], spike_indices[i])
+
+        return sample_spikes
+
+    def run_extract_waveforms(self):
+        self.sorting_analyzer.compute("waveforms")
```
Comment (alejoe91) [inline src/spikeinterface/core/waveforms_extractor_backwards_compatibility.py:272]: Remove. The `MockWaveformExtractor` should not add new computations

**Review by alejoe91 — CHANGES_REQUESTED** [review]
Added comments after discussion

File: src/spikeinterface/core/analyzer_extension_core.py [hunk 4]
```diff
@@ -319,6 +319,9 @@ def _compute_and_append(self, operators):
 
         # spikes = self.sorting_analyzer.sorting.to_spike_vector()
         # some_spikes = spikes[self.sorting_analyzer.random_spikes_indices]
+        assert self.sorting_analyzer.has_extension(
+            "random_spikes"
+        ), "compute templates requires the random_spikes extension. You can run WaveformExtractor.get_random_spikes()"
```
Comment (alejoe91) [inline src/spikeinterface/core/analyzer_extension_core.py:324]:
```suggestion
        assert self.sorting_analyzer.has_extension(
            "random_spikes"
        ), "compute templates requires the random_spikes extension. You can run sorting_analyzer.get_random_spikes()"
```

**Review by samuelgarcia — APPROVED** [review]

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**